cloudinit/config/cc_grub_dpkg.py

```diff
@@ -62,22 +86,8 @@ def handle(name, cfg, _cloud, log, _args):
     idevs_empty = util.get_cfg_option_str(
         mycfg, "grub-pc/install_devices_empty", None)
 
-    if ((os.path.exists("/dev/sda1") and not os.path.exists("/dev/sda")) or
-       (os.path.exists("/dev/xvda1") and not os.path.exists("/dev/xvda"))):
-        if idevs is None:
-            idevs = ""
-        if idevs_empty is None:
-            idevs_empty = "true"
-    else:
-        if idevs_empty is None:
-            idevs_empty = "false"
-        if idevs is None:
-            idevs = "/dev/sda"
-            for dev in ("/dev/sda", "/dev/vda", "/dev/xvda",
-                        "/dev/sda1", "/dev/vda1", "/dev/xvda1"):
-                if os.path.exists(dev):
-                    idevs = dev
-                    break
+    if idevs is None:
```

Comment:
If `idevs` is not `None` but `idevs_empty` _is_, then we will attempt to pass `grub-pc grub-pc/install_devices_empty boolean None\n` into debconf-set-selections below (which I assume won't go well :-).

If `fetch_idevs` were to change from returning `(str, str)` to returning `(Optional[str], Optional[str])`[0], that would gives it the same return types as the `util.get_cfg_option_str(..., None)` calls above. This in turn could mean that the defaulting logic could move out of `fetch_idevs` and into this function (or perhaps a helper function).

Your refactor has simplified this function enough that it looks like it might be reasonable to write unit tests for it (which would hopefully catch/have caught the issue I spotted here). Would you be able to add some?

---

[0] If `Optional[str]` doesn't mean anything to you: it's the Python type annotation way of expressing "either a str or None", see [the `typing.Optional` docs](https://docs.python.org/3/library/typing.html#typing.Optional). Other languages call this Maybe or Option: https://en.wikipedia.org/wiki/Option_type

Reply:
Great catch. I moved the logic for deciding `idevs_empty` out of `fetch_idevs` and back into `handle`, and it now conforms to how the previous implemented handled it. 

I have added a new unit test class which tests `handle()` as you requested, and it covers the test matrix of all idevs and idevs_empty configuration pairs.